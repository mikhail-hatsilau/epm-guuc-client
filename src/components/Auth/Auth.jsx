import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

import Api from '../../api';

import styles from './Auth.module.scss';

export default class Auth extends Component {
    static propTypes = {
        onLoggedIn: PropTypes.func
    };

    static validatePassword = (password) => {
        const MIN_PASSWORD_LENGTH = 6;
        const passLength = password.length;
        if (passLength < MIN_PASSWORD_LENGTH) {
            if (passLength === 0) {
                return 'Field can not be empty'
            }

            return `Password should be more then ${MIN_PASSWORD_LENGTH} symbols`
        }
    };

    static validateUsername = (username) => {

    };


    constructor(props) {
        super(props);
        this.state = {
            username: '',
            password: '',
            invalidUsername: false,
            invalidPassword: false
        }
    }

    componentDidMount() {
        Api.fetchUser(1).then((res) => {
            console.log(res.user);
        })
    }

    handleUsernameChange = (event) => {
        this.setState({
            username: event.target.value
        });
    };

    handlePasswordChange = (event) => {
        const password = event.target.value;
        const invalidPasswordMessage = Auth.validatePassword(password);

        this.setState({
            invalidPassword: !!invalidPasswordMessage,
            invalidPasswordMessage,
            password
        });
    };

    handleSubmit = () => {
        const { invalidPassword, invalidUsername } = this.state;

        if (!invalidPassword && !invalidUsername) {
            this.props.onLoggedIn();
        }
    };

    render() {
        const { invalidPasswordMessage } = this.state;
        const passwordClassNames = classnames({
            [styles.invalid]: invalidPasswordMessage
        });
        return (
            <div className={ styles.authComponent }>
                <div>
                    <label htmlFor="username">Username:</label>
                    <input id="username" type="text" onChange={ this.handleUsernameChange } />
                </div>
                <div>
                    <label htmlFor="password">Password:</label>
                    <input id="password" type="password" className={ passwordClassNames } onChange={ this.handlePasswordChange } />
                    { invalidPasswordMessage && <div className={ styles.error }>{ invalidPasswordMessage }</div> }
                </div>
                <div className={ styles.submit }>
                    <button onClick={ this.handleSubmit }>Sign in</button>
                </div>
                <div></div>
            </div>
        )
    }
}

