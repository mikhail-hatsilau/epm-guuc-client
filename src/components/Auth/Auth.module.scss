.authComponent {
}

.password {

}

.invalid {
  border-color: red;
}