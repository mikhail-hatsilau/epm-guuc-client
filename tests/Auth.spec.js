import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Auth from '../src/components/Auth';

jest.mock('../src/api', () => (
    {
        fetchUser: jest.fn((id) => Promise.resolve({ user: { username: 'test' } }))
    }
));

describe('Auth', () => {
    let component;

    beforeEach(() => {
        component = renderer.create(<Auth />);
    });

    it('should be rendered', () => {
        const wrapper = shallow(<Auth />);
        expect(wrapper.find('.authComponent')).toBeTruthy();
    });

    it('renders correctly', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should handle password validation', () => {
        const instance = component.getInstance();

        expect(component.toJSON()).toMatchSnapshot();

        instance.handlePasswordChange({
            target: {
                value: '1234'
            }
        });

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should call onLoggedIn callback when form is submitted', () => {
        const onLoggedIn = jest.fn();
        component = shallow(<Auth onLoggedIn={ onLoggedIn }/>);

        component.setState({
            invalidUsername: false,
            invalidPassword: false
        });

        component.find('button').simulate('click');

        expect(onLoggedIn).toBeCalled();
    });

    it('should fetch user data', () => {
        const apiModule = require('../src/api');
        apiModule.fetchUser.mockClear();
        component = renderer.create(<Auth />);

        expect(apiModule.fetchUser.mock.calls).toHaveLength(1);
        expect(apiModule.fetchUser).toHaveBeenCalledWith(1);
    });
});