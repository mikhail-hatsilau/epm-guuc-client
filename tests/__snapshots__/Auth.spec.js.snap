exports[`Auth renders correctly 1`] = `
<div
  className="authComponent">
  <div>
    <label
      htmlFor="username">
      Username:
    </label>
    <input
      id="username"
      onChange={[Function]}
      type="text" />
  </div>
  <div>
    <label
      htmlFor="password">
      Password:
    </label>
    <input
      className=""
      id="password"
      onChange={[Function]}
      type="password" />
  </div>
  <div
    className="submit">
    <button
      onClick={[Function]}>
      Sign in
    </button>
  </div>
  <div />
</div>
`;

exports[`Auth should handle password validation 1`] = `
<div
  className="authComponent">
  <div>
    <label
      htmlFor="username">
      Username:
    </label>
    <input
      id="username"
      onChange={[Function]}
      type="text" />
  </div>
  <div>
    <label
      htmlFor="password">
      Password:
    </label>
    <input
      className=""
      id="password"
      onChange={[Function]}
      type="password" />
  </div>
  <div
    className="submit">
    <button
      onClick={[Function]}>
      Sign in
    </button>
  </div>
  <div />
</div>
`;

exports[`Auth should handle password validation 2`] = `
<div
  className="authComponent">
  <div>
    <label
      htmlFor="username">
      Username:
    </label>
    <input
      id="username"
      onChange={[Function]}
      type="text" />
  </div>
  <div>
    <label
      htmlFor="password">
      Password:
    </label>
    <input
      className="invalid"
      id="password"
      onChange={[Function]}
      type="password" />
    <div
      className="error">
      Password should be more then 6 symbols
    </div>
  </div>
  <div
    className="submit">
    <button
      onClick={[Function]}>
      Sign in
    </button>
  </div>
  <div />
</div>
`;
